File: src4/main/java/com/regenerationforrged/data/worldgen/preset/settings/ContinentType.java
package com.regenerationforrged.data.worldgen.preset.settings;

import com.mojang.serialization.Codec;

import net.minecraft.util.StringRepresentable;
import com.regenerationforrged.world.worldgen.GeneratorContext;
import com.regenerationforrged.world.worldgen.cell.continent.Continent;
import com.regenerationforrged.world.worldgen.cell.continent.advanced.AdvancedContinentGenerator;
import com.regenerationforrged.world.worldgen.cell.continent.fancy.FancyContinentGenerator;
import com.regenerationforrged.world.worldgen.cell.continent.infinite.InfiniteContinentGenerator;
import com.regenerationforrged.world.worldgen.cell.continent.simple.MultiContinentGenerator;
import com.regenerationforrged.world.worldgen.cell.continent.simple.SingleContinentGenerator;
import com.regenerationforrged.world.worldgen.util.Seed;

public static class ContinentType {

    public static final Codec<Continent> CODEC =
            RecordCodecBuilder.create(instance -> instance.group(
                    ContinentType.CODEC.fieldOf("continentType").forGetter(o -> o.continentType),
                    DistanceFunction.CODEC
                            .optionalFieldOf("continentShape", DistanceFunction.EUCLIDEAN)
                            .forGetter(o -> o.continentShape),
                    Codec.INT.fieldOf("continentScale").forGetter(o -> o.continentScale),
                    Codec.FLOAT.fieldOf("continentJitter").forGetter(o -> o.continentJitter),
                    Codec.FLOAT.optionalFieldOf("continentSkipping", 0.25F).forGetter(o -> o.continentSkipping),
                    Codec.FLOAT.optionalFieldOf("continentSizeVariance", 0.25F).forGetter(o -> o.continentSizeVariance),
                    Codec.INT.optionalFieldOf("continentNoiseOctaves", 5).forGetter(o -> o.continentNoiseOctaves),
                    Codec.FLOAT.optionalFieldOf("continentNoiseGain", 0.26F).forGetter(o -> o.continentNoiseGain),
                    Codec.FLOAT.optionalFieldOf("continentNoiseLacunarity", 4.33F).forGetter(o -> o.continentNoiseLacunarity)
            ).apply(instance, Continent::new));

    public final ContinentType continentType;
    public final DistanceFunction continentShape;
    public final int continentScale;
    public final float continentJitter;
    public final float continentSkipping;
    public final float continentSizeVariance;
    public final int continentNoiseOctaves;
    public final float continentNoiseGain;
    public final float continentNoiseLacunarity;

    public ContinentType(ContinentType type, DistanceFunction shape, int scale,
                     float jitter, float skipping, float sizeVariance,
                     int octaves, float gain, float lacunarity) {

        this.continentType = type;
        this.continentShape = shape;
        this.continentScale = scale;
        this.continentJitter = jitter;
        this.continentSkipping = skipping;
        this.continentSizeVariance = sizeVariance;
        this.continentNoiseOctaves = octaves;
        this.continentNoiseGain = gain;
        this.continentNoiseLacunarity = lacunarity;
    }

    public Continent copy() {
        return new Continent(
                continentType,
                continentShape,
                continentScale,
                continentJitter,
                continentSkipping,
                continentSizeVariance,
                continentNoiseOctaves,
                continentNoiseGain,
                continentNoiseLacunarity
        );
    }
}

File: src2/main/kotlin/com/mbz/data/worldgen/preset/settings/ContinentType.kt
package com.mbz.data.worldgen.preset.settings

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.util.StringRepresentable
import com.mbz.world.worldgen.GeneratorContext
import com.mbz.world.worldgen.cell.continent.*
import com.mbz.world.worldgen.cell.continent.advanced.AdvancedContinentGenerator
import com.mbz.world.worldgen.cell.continent.fancy.FancyContinentGenerator
import com.mbz.world.worldgen.cell.continent.infinite.InfiniteContinentGenerator
import com.mbz.world.worldgen.cell.continent.simple.*
import com.mbz.world.worldgen.noise.function.DistanceFunction

data class Continent(
    var continentType: ContinentType = ContinentType.MULTI,
    var continentShape: DistanceFunction = DistanceFunction.EUCLIDEAN,
    var continentScale: Int = 500,
    var continentJitter: Float = 0.45f,
    var continentSkipping: Float = 0.25f,
    var continentSizeVariance: Float = 0.25f,
    var continentNoiseOctaves: Int = 5,
    var continentNoiseGain: Float = 0.26f,
    var continentNoiseLacunarity: Float = 4.33f
) {
    companion object {
        @JvmField
        val CODEC: Codec<Continent> = RecordCodecBuilder.create { i ->
            i.group(
                ContinentType.CODEC.fieldOf("continentType").forGetter { it.continentType },
                DistanceFunction.CODEC.optionalFieldOf("continentShape", DistanceFunction.EUCLIDEAN).forGetter { it.continentShape },
                Codec.INT.fieldOf("continentScale").forGetter { it.continentScale },
                Codec.FLOAT.fieldOf("continentJitter").forGetter { it.continentJitter },
                Codec.FLOAT.optionalFieldOf("continentSkipping", 0.25f).forGetter { it.continentSkipping },
                Codec.FLOAT.optionalFieldOf("continentSizeVariance", 0.25f).forGetter { it.continentSizeVariance },
                Codec.INT.optionalFieldOf("continentNoiseOctaves", 5).forGetter { it.continentNoiseOctaves },
                Codec.FLOAT.optionalFieldOf("continentNoiseGain", 0.26f).forGetter { it.continentNoiseGain },
                Codec.FLOAT.optionalFieldOf("continentNoiseLacunarity", 4.33f).forGetter { it.continentNoiseLacunarity }
            ).apply(i, ::Continent)
        }
    }
}

enum class ContinentType(val id: String) : StringRepresentable {
    SINGLE("SINGLE"),
    MULTI("MULTI"),
    FANCY("FANCY"),
    INFINITE("INFINITE"),
    ADVANCED("ADVANCED");

    override fun getSerializedName() = id

    companion object {
        @JvmField
        val CODEC: Codec<ContinentType> = StringRepresentable.fromEnum { values() }
    }
}
